File: 7-reverse-integer/7-reverse-integer.java
class Solution {
    public int reverse(int x) {
        long res = 0;
        while(x != 0) {
            if(res > 0 && res > (Integer.MAX_VALUE - x % 10) / 10) return 0;
            if(res < 0 && res < (Integer.MIN_VALUE - x % 10) / 10) return 0;
            res = res* 10 + x % 10;
            x /= 10;
        }
        return (int) res;
    }
}

File: 7-reverse-integer/7-reverse-integer.py
class Solution:
    def reverse(self, x: int) -> int:
        neg = 1
        if x < 0: neg = -1
        res = 0
        lower, upper = -2**31, 2**31 - 1
        x = abs(x)
        while x != 0:
            if neg == 1 and res > (upper - x % 10) / 10:
                return 0
            if neg == -1 and res > (upper - x % 10 + 1) / 10:
                return 0
            
            res = res * 10 + x % 10
            x //= 10
        return res * neg
            
        